core/src/test/java/io/grpc/internal/RetriableStreamTest.java

```diff
@@ -0,0 +1,557 @@
+/*
+ * Copyright 2017, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.inOrder;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import io.grpc.Compressor;
+import io.grpc.DecompressorRegistry;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.MethodDescriptor.MethodType;
+import io.grpc.Status;
+import io.grpc.StringMarshaller;
+import java.io.InputStream;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.ArgumentCaptor;
+import org.mockito.InOrder;
+
+/**
+ * Unit tests for {@link RetriableStream}.
+ */
+@RunWith(JUnit4.class)
+public class RetriableStreamTest {
+  private static final String CANCELLED_BECAUSE_COMMITTED =
+      "Stream thrown away because RetriableStream committed";
+  private static final String AUTHORITY = "fakeAuthority";
+  private static final Compressor COMPRESSOR = mock(Compressor.class);
+  private static final DecompressorRegistry DECOMPRESSOR_REGISTRY =
+      DecompressorRegistry.getDefaultInstance();
+  private static final int MAX_INBOUND_MESSAGE_SIZE = 1234;
+  private static final int MAX_OUTNBOUND_MESSAGE_SIZE = 5678;
+
+  private final MethodDescriptor<String, String> method =
+      MethodDescriptor.<String, String>newBuilder().setType(MethodType.BIDI_STREAMING)
+          .setFullMethodName(MethodDescriptor.generateFullMethodName("service_foo", "method_bar"))
+          .setRequestMarshaller(new StringMarshaller())
+          .setResponseMarshaller(new StringMarshaller()).build();
+  private final ClientStreamListener masterListener = mock(ClientStreamListener.class);
+
+  // TODO(zdapeng): for mockito 2.7+: instead of spy(an_impl),
+  // use mock(RetriableStream.class, withSettings().useConstructorArgs(method))
+  private final RetriableStream<String> retriableStream = spy(new RetriableStream<String>(method) {
```

Comment:
Whoa. That's scary. We should not use a spy on the code under test. Spy does evilness that means I wouldn't trust results.